File: lib/models/book.dart
class Book {
  final String id;
  final String title;
  final String author;
  final String coverImage;
  final String description;
  final double rating;
  final List<String> tags;
  final List<Chapter> chapters;
  bool isHidden;
  final bool isFromFile;

  Book({
    required this.id,
    required this.title,
    required this.author,
    required this.coverImage,
    required this.description,
    required this.rating,
    required this.tags,
    required this.chapters,
    this.isHidden = false,
    this.isFromFile = false,
  });

  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: (json['bookId'] ?? json['id'] ?? '').toString(),
      title: json['title'] ?? '',
      author: json['author'] ?? '',
      coverImage: json['imageUrl'] ?? json['coverImage'] ?? '',
      description: json['description'] ?? '',
      rating: (json['averageScore'] ?? json['rating'] as num?)?.toDouble() ?? 0.0,
      tags: (json['tags'] as List<dynamic>?)?.map((e) {
            if (e is Map) return e['tagName']?.toString() ?? '';
            return e.toString();
          }).toList() ??
          [],
      chapters: ((json['volumes'] ?? json['chapters']) as List<dynamic>?)
              ?.map((e) => Chapter.fromJson(e))
              .toList() ??
          [],
      isHidden: json['isHidden'] ?? false,
      isFromFile: json['isCoverFromFile'] ?? false,
    );
  }
}

class Chapter {
  final String id;
  final String title;
  final String content;
  final int chapterNumber;

  Chapter({
    required this.id,
    required this.title,
    required this.content,
    required this.chapterNumber,
  });

  factory Chapter.fromJson(Map<String, dynamic> json) {
    return Chapter(
      id: (json['volumeId'] ?? json['id'] ?? '').toString(),
      title: json['title'] ?? '',
      content: json['content'] ?? '',
      chapterNumber: json['volumeNo'] ?? json['chapterNumber'] ?? 0,
    );
  }
}

File: BookAppAPI/Models/Book.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace BookAppAPI.Models
{
    public class Book
    {
        [Key]
        public int BookId { get; set; }

        [Required]
        [StringLength(200)]
        public string Title { get; set; } = string.Empty;

        [Required]
        [StringLength(100)]
        public string Author { get; set; } = string.Empty;

        public string? Description { get; set; }

        [StringLength(500)]
        public string? ImageUrl { get; set; }

        [Required]
        [StringLength(10)]
        public string ContentType { get; set; } = "file";

        [StringLength(500)]
        public string? FileUrl { get; set; }

        public int CategoryId { get; set; }

        public int TotalVotes { get; set; } = 0;
        public int TotalScore { get; set; } = 0;
        public bool IsCoverFromFile { get; set; } = false;
        public bool IsHidden { get; set; } = false;
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

        // Navigation properties
        [ForeignKey("CategoryId")]
        public Category Category { get; set; } = null!;
        public ICollection<BookTag> BookTags { get; set; } = new List<BookTag>();
        public ICollection<Volume> Volumes { get; set; } = new List<Volume>();
        public ICollection<Rating> Ratings { get; set; } = new List<Rating>();
        public ICollection<Comment> Comments { get; set; } = new List<Comment>();
        public ICollection<Favorite> Favorites { get; set; } = new List<Favorite>();
        public ICollection<ReadingHistory> ReadingHistories { get; set; } = new List<ReadingHistory>();

        [NotMapped]
        public double AverageScore => TotalVotes > 0 ? (double)TotalScore / TotalVotes : 0;
    }
}
